bluemira/geometry/parameterisations.py

```diff
@@ -1681,41 +1683,75 @@ def _make_tapered_inner_leg(
             Vertical position of top of limb [m]
         z2:
             Vertical position of bottom of limb [m]
+        r_min:
+            Minimum radius of curvature [m]
         axis:
             [x,y,z] vector normal to plane of parameterisation
 
         Returns
         -------
         CAD Wire of the geometry
         """
-        # Bottom straight section
-        p1 = [x_mid, 0, -z_in]
-        p2 = [x_mid, 0, z2]
-        bot_straight = make_polygon([p2, p1], label="inner_limb_mid_down")
-
-        # Curved taper radius
-        x_t = x_mid - x_in
-        alpha = np.arctan(z_in / (x_t))
-        theta_t = np.pi - 2 * alpha
-        r_taper = z_in / np.sin(theta_t)
+        # Pre-calculations necessary
+        if not (z2 < -z_in < 0 < z_in < z1):
+            raise ValueError(
+                "the straight-curve transition point z_in must lie between z1 and z2."
+            )
+        # epsilon must be larger than np.finfo().eps because of quirks of floats >2.0.
+        eps = np.finfo(2 * np.pi).resolution
+        dx = x_mid - x_in
+        expression_to_find_root_for = lambda x: np.sin(x) + (np.cos(x) - 1) * (z_in / dx)
+        theta = optimize.bisect(expression_to_find_root_for, eps, 2 * np.pi - eps)
```

Comment:
Right, so before I just reacted allergically to the use of scipy optimisation, because I do not think it is a good or robust toolkit. minimize or optimize, who cares.

Now, I've looked at where it is being used, I'm wondering why we need anything at all, other than numpy and trigonometry.

Bear with me, I have not slept properly in weeks, and I have yet to have coffee number 3. Please check the below, and tell me what I am missing...

The expression is of the form:
sin(x)+a(cos(x)-1)=0
which has an obvious trivial root of 0, and n slightly less obvious roots, which have an exact form:

```python
import numpy as np
from scipy.optimize import bisect

eps = np.finfo(2 * np.pi).resolution

A = 2.46  # random constant != 0

def func(x):
    return np.sin(x) + A* (np.cos(x) - 1)

theta = bisect(func, eps, 2 * np.pi - eps)

# First non-trivial root - just in case of arctan shenanigans
theta_exact = np.arctan2((2*A)/(A**2+1), (A**2-1)/(A**2+1))
# Alternate form and nth root
n = 1
theta_exact1 = 2*(np.arctan(1/A)+n*np.pi)
n = 2
theta_exact2 = 2*(np.arctan(1/A)+n*np.pi)

print(func(theta))        # 4.765077221691172e-13
print(func(theta_exact))  # 3.3306690738754696e-16
print(func(theta_exact1)) # 2.220446049250313e-16
print(func(theta_exact2)) # 6.661338147750939e-16
```

Reply:
Thank you, that's a great point, I forgot to do the 1-cos(2x) = 2sin²(x) substitution so I just forgot this can be simplified & solved analytically. I'll use the exact solution then.